src/tfr_sensor/src/fiducial_odom_publisher.cpp

```diff
@@ -0,0 +1,277 @@
+                        relative_bin_transform.rotation.z,
+                        relative_bin_transform.rotation.w);
+                 //take a difference of the two transforms to find the
+                //odom_camera transform
+                tf2::Transform p_0{};
+                tf2::convert(processed_pose.pose, p_0);
+                tf2::Transform p_1{};
+                tf2::convert(relative_bin_transform, p_1);
+                //get the difference between the two transforms
+                auto difference = p_1.inverseTimes(p_0);
+                geometry_msgs::Transform relative_transform;
+                relative_transform = tf2::toMsg(difference);
+                geometry_msgs::PoseStamped relative_pose;
+                relative_pose.header.stamp = ros::Time::now();
+                relative_pose.header.frame_id = camera_frame;
+                relative_pose.pose.position.x = -relative_transform.translation.x;
+                relative_pose.pose.position.y = -relative_transform.translation.y;
+                relative_pose.pose.position.z = -relative_transform.translation.z;
+                relative_pose.pose.orientation = relative_transform.rotation;
+                ROS_INFO("relative data %s %f %f %f %f %f %f %f",
+                        relative_pose.header.frame_id.c_str(),
+                        relative_pose.pose.position.x,
+                        relative_pose.pose.position.y,
+                        relative_pose.pose.position.z,
+                        relative_pose.pose.orientation.x,
+                        relative_pose.pose.orientation.y,
+                        relative_pose.pose.orientation.z,
+                        relative_pose.pose.orientation.w);
+ 
+                // 1. handle transforms for tf
+//                geometry_msgs::TransformStamped transform;
+//                transform.header.stamp = ros::Time::now();
+//                transform.header.frame_id = odometry_frame;
+//                transform.child_frame_id = footprint_frame;
+//                transform.transform = relative_transform;
+//                broadcaster.sendTransform(transform);
+
+                // 2. handle odometry data
+                nav_msgs::Odometry odom;
+                odom.header.frame_id = odometry_frame;
+                odom.header.stamp = ros::Time::now();
+                odom.child_frame_id = footprint_frame;
+                //get our pose and fudge some covariances
+                odom.pose.pose = relative_pose.pose;
+                odom.pose.covariance = {  5e-3,   0,   0,   0,   0,   0,
+                                             0,5e-3,   0,   0,   0,   0,
+                                             0,   0,5e-3,   0,   0,   0,
+                                             0,   0,   0,5e-3,   0,   0,
+                                             0,   0,   0,   0,5e-3,   0,
+                                             0,   0,   0,   0,   0,5e-3};
+
```

Comment:
Is this covariance matrix made up?  Or do these numbers come from somewhere?

Reply:
It's made up. In fact all of our covariance matrices are made up from completely arbitrary numbers (none of our device drivers shipped with them, I had to add them manually). The place where they matter is in `robot_localization`, it uses them as part of it's state estimation algorithm. Populating these with semi reasonable values is a `//TODO` for our sensor fusion team, as part of tuning the kahlman filter.